Sources/GRPCHTTP2TransportNIOPosix/HTTP2ServerTransport+Posix.swift

```diff
@@ -21,13 +21,73 @@ import NIOExtras
 import NIOPosix
 
 extension HTTP2ServerTransport {
+  /// A NIOPosix-backed implementation of a server transport.
   @available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *)
   public struct Posix: ServerTransport {
     private let address: GRPCHTTP2Core.SocketAddress
     private let config: Config
     private let eventLoopGroup: MultiThreadedEventLoopGroup
     private let serverQuiescingHelper: ServerQuiescingHelper
 
+    private enum State {
+      case idle(IdleState)
+      case listening(ListeningState)
+      case closed(ClosedState)
+
+      struct IdleState {
+        var listeningAddressPromise: EventLoopPromise<NIOCore.SocketAddress>
+        var noLongerListeningPromise: EventLoopPromise<NIOCore.SocketAddress>
+      }
+
+      struct ListeningState {
+        var listeningAddressFuture: EventLoopFuture<NIOCore.SocketAddress>
+        var noLongerListeningPromise: EventLoopPromise<NIOCore.SocketAddress>
+
+        init(idleState: IdleState) {
+          self.listeningAddressFuture = idleState.listeningAddressPromise.futureResult
+          self.noLongerListeningPromise = idleState.noLongerListeningPromise
+        }
+      }
+
+      struct ClosedState {
+        var noLongerListeningFuture: EventLoopFuture<NIOCore.SocketAddress>
+
+        init(idleState: IdleState) {
+          self.noLongerListeningFuture = idleState.noLongerListeningPromise.futureResult
+        }
+
+        init(listeningState: ListeningState) {
+          self.noLongerListeningFuture = listeningState.noLongerListeningPromise.futureResult
+        }
+      }
+    }
+    private let listeningAddressState: _LockedValueBox<State>
+
+    /// The listening address for this server transport.
+    ///
+    /// It is an `async` property because it will only return once the address has been successfully bound.
+    ///
+    /// - Throws: A runtime error will be thrown if the address could not be bound or is not bound any
+    /// longer, because the transport isn't listening anymore.
+    public var listeningAddress: NIOCore.SocketAddress {
```

Comment:
We still want this to be a gRPC socket address, we need to convert from NIOs socket address where applicable